### ISSUE TITLE
Runtime error when blocking in Scala.js

### ISSUE THREAD

**Issue by durban** [typelevel/cats-effect#43]
`fs.Task` doesn't allow calling `unsafeRun()` when compiling for Scala.js, it causes a *compile-time* error. As far as I can tell, `cats.effect.IO` allows calling `unsafeRunSync()`, and it will produce a *runtime* error on Scala.js. Would it be possible to have a compile-time error here too? (I'm sure it's technically possible to implement, so my question is more about why was it implemented this way ... what is the upside to this implementation?)

**Comment by alexandru** [comment]
I'm personally all for banning `unsafeRunSync` altogether, even from the JVM side, but @djspiewak is going to come with a story about having some `IO` references guaranteed to not trigger async boundaries on `unsafeRunSync`, you'll believe him and then go back to suffering from runtime errors 😜

**Comment by djspiewak** [comment]
There are several upsides to this implementation, but looping back to some prerequisite understanding.

`unsafeRunSync` is not inherently broken on JavaScript.  `unsafeRunSync` *on an `IO` which contains `async` components* is inherently broken.  And arguably not just on JavaScript, since the JVM only "gets away with it" because it blocks a thread (as a sidebar, I know how to avoid thread blocking, but it requires something more powerful than `ExecutionContext` and probably isn't worth it in the end).  But `unsafeRunSync` on a fully synchronous `IO` is completely fine.  `IO(println("hi")).unsafeRunSync()` has no problems on either JavaScript or the JVM.

So the problem we have here is that this function fails because of dynamic state that isn't reflected in the types.  `IO` is two monads mashed together, and we don't distinguish between those cases, so we don't have any way to forbid `unsafeRunSync` on asynchronous `IO`s while allowing it on synchronous ones.  Which is why we have to choose between two unappealing alternatives: over-restrict and throw the whole thing out, or under-restrict and produce runtime errors when people do something nonsensical.  I chose the latter.

There are two strong benefits for having `unsafeRunSync`.  The first is testing.  It is much, much more convenient in a test to just have a synchronous, `IO[A] => A` function.  And there the runtime errors don't matter really at all, since it's only going to break the test and it will (very correctly) signal that you need to fix something.  The second benefit is parametricity.  The type signature `IO[A] => A` guarantees that `IO` interprets itself; it is intrinsically strong enough to represent the richness of its free algebra.  The only other alternative would be either `IO[A] => (Either[Throwable, A] => Unit) => Unit`, which is very difficult to digest and severely obfuscates the point, or `IO[A] => Future[A]`, which very incorrectly implies that `IO` interprets to `Future` (and furthermore implies that `Future` is powerful enough to embed `IO`, which it is not).

The `runAsync` function allows us to have the best of both worlds.  An `IO` returned from `runAsync` is guaranteed to be safe to call `unsafeRunSync`.  On *every* platform.  This is how the laws are defined, and it is generally assumed that most production use will fall into this pattern.

I can't prevent users from constructing nonsense.  `IO` is particularly risky in that regard since there are so many unsafe functions.  If we try to go down the route of over-constraining things that are unsafe in certain applications, `IO` wouldn't exist in the first place.  So `unsafeRunSync` remains, and it throws an exception on Scala.js if it hits an async boundary.  Don't do that.  Either be *really* sure that the `IO` you're running consists of solely synchronous segments, or guarantee it by calling `runAsync` first.

> you'll believe him and then go back to suffering from runtime errors

It's not a matter of belief.  If you use the `IO` returned from `runAsync` is guaranteed not to produce runtime errors when synchronously run.  That's like saying "you'll believe him when he says the earth is round."  No faith required; it's just a fact.

**Comment by alexandru** [comment]
@djspiewak that was my way of saying you've got good arguments for it, with a (bad?) joke. Hope I haven't offended you.

**Comment by durban** [comment]
@djspiewak Thank you for the detailed explanation! I think I'm starting to understand. However, I have a few more questions, if you don't mind. (I'm not trying to be difficult, I'd just like to understand the tradeoffs involved. BTW feel free to direct me to some documentation, or something, if the answers alre already written down somewhere.)

You said that `IO` is two monads together, and "we don't distinguish between those cases". My question is: why don't we? Could we have something like `MayBeAsyncIO` and `SyncIO extends MayBeAsyncIO` (with less terrible names), and only `SyncIO` having `unsafeRunSync`? (Or something equivalent, but without subtyping, if that causes problems?)

Just to make sure I understand: the 2 monads together are (1) something like `Suspendable` in FS2, which can be run with `unsafeRunSync`; and (2) `cats.effect.Async`. Is (1) the same as `cats.effect.Sync`? Is (1) the same as `cats.Eval`? (Or `Eval` can't be used for effect capture? I seem to remember something like this...)

(I still have to digest how `runAsync` works, and how can it guarantee that it can be run synchronously.)

Thanks!

**Comment by alexandru** [comment]
@durban 

> You said that IO is two monads together, and "we don't distinguish between those cases". My question is: why don't we? Could we have something like MayBeAsyncIO and SyncIO extends MayBeAsyncIO (with less terrible names), and only SyncIO having unsafeRunSync? (Or something equivalent, but without subtyping, if that causes problems?)

We already have `cats.Eval` in `cats-core`, which is basically the `SyncIO` that you're talking about. And you can abstract between them using the available type-classes. I also used this exact rationale in Monix, when I designed `Task` and `Coeval`.

> (1) something like Suspendable in FS2, which can be run with unsafeRunSync; Is (1) the same as cats.effect.Sync? 

Yes, `Sync` is `Suspendable` from FS2.

> Or Eval can't be used for effect capture? I seem to remember something like this...)

Yes it can, although the current version does not implement `MonadError`, which means you can't do error handling with it, but the next version will do that as well.

> (I still have to digest how runAsync works, and how can it guarantee that it can be run synchronously.)

`runAsync` does nothing more but to trigger `unsafeRunAsync` internally, but the generated `IO` instance does not wait for anything on evaluation, so it doesn't need to cross any async boundaries.

**Comment by milessabin** [comment]
> IO is two monads mashed together, and we don't distinguish between those cases, so we don't have any way to forbid unsafeRunSync on asynchronous IOs while allowing it on synchronous ones. Which is why we have to choose between two unappealing alternatives: over-restrict and throw the whole thing out, or under-restrict and produce runtime errors when people do something nonsensical. I chose the latter.

Is there a third option? Distinguish between sync and async in the type?

**Comment by djspiewak** [comment]
> Is there a third option? Distinguish between sync and async in the type?

This is basically the same as @durban's question, and it's an interesting one.

There are two reasons we don't do this.  The first is convenient typing and convenient typing.

> The more interesting your types get, the less fun it is to write them down! – Benjamin Pierce

(ideally read the above in a Ken Burns style, with a fiddle playing in the background panning over a sepia photo)

Anyway, we have to write the `IO` type constructor a lot, and it just becomes more annoying if we have to constantly write `IO[Sync, A]` or `IO[Async, A]`, or carry around a more complex proof mechanism.  The cost from a usability standpoint is pretty high for a very small gain, especially since we *still* would be over-approximating.  Any `IO` that we couldn't *definitely* prove is fully synchronous would need to be assumed asynchronous, and that's still going to over-approximate the constraint.

The second reason not to do this is flatMap.  Because we need to frame a type constructor solely in terms of the variable `A`, we would need to either define multiple `Monad`s over `IO`, or we would need to assume that any `flatMap` on `IO` results in an asynchronous `IO`, and only works *over* asynchronous `IO`.  This also implies subsumption (`Sync <: Async` in our phantom constraint).

Basically, it ends up getting pretty messy and being a very lossy constraint system, ultimately paying a significant penalty for very little gain.

**Comment by milessabin** [comment]
OK, hypothetically, could this be made to work out more smoothly in, eg. Idris? If so, what are the minimal language extensions required to make this tolerable in Scala?

**Comment by djspiewak** [comment]
@milessabin It could definitely be made more smooth in Idris in the sense that everything is more smooth in Idris, but I think a lot of the same problems remain.  There's still extra typing, you still have to carry extra proofs around (though it would be much easier to do so), and you would still need to aggressively subsume up to the least-constrained case when passing through bind.

Ultimately, `Monad` just isn't that powerful of an interface, because of the type constructor constraint.  If we had some sort of `CMonad[C[_], F[_]]` which allowed carrying an arbitrary proof around (think: `CFunctor[Order, Set]`), you might be able to do this conveniently, but experiments of that nature have generally proven unsatisfactory (think: `TransLift` in cats), and not just because of Scala-related reasons.

**Comment by milessabin** [comment]
Would `SyncIO[A] <: IO[A]` (as an alternative to a phantom type) get us anywhere? With `flatMap` suitably refined for the subtype?

**Comment by djspiewak** [comment]
@milessabin How would you write the type of `flatMap` in such a way that it retains specificity of `SyncIO`, enabling refined override (with *covariant* refinement in the parameter type) while falling back to `IO` and remaining compatible with the generic `Monad` abstraction?  This is the exact use-case that motivated `CanBuildFrom`.  :-(

**Comment by milessabin** [comment]
If `flatMap` isn't defined directly on `IO`/`SyncIO` then that doesn't have to be a problem.

**Comment by djspiewak** [comment]
So you would define it externally and cast? It's a free monad, so flat map uses the internal structure

**Comment by alexandru** [comment]
This discussion is really interesting, but I agree with @djspiewak because exposing more complex types to the user has a cost, the extra difficulty has to be worth it and pragmatic APIs with a lower learning cost tend to win.

And in this case it's much easier to either remove `unsafeRunSync` completely or to expose it and push that responsibility to the user. Its usage shouldn't be that common anyway. And if we need a `SyncIO`, that role can be fulfilled by `cats.Eval`.

**Comment by milessabin** [comment]
We could provide the implementations directly on the `FlatMap` instances for `IO` and `SyncIO` where they could have access to internal structure without any need to cast.

**Comment by djspiewak** [comment]
@milessabin If `SyncIO <: IO`, then `SyncIO` contains all of the potential structure that `IO` has.  Even if there isn't a cast, there would *at least* be an unsafe pattern match.  The `Sync` part of the constraint would be literally phantom.

> And in this case it's much easier to either remove unsafeRunSync completely or to expose it and push that responsibility to the user. Its usage shouldn't be that common anyway.

Its usage in tests is extremely common.

**Comment by alexandru** [comment]
> > And in this case it's much easier to either remove unsafeRunSync completely or to expose it and push that responsibility to the user. Its usage shouldn't be that common anyway.
>
> Its usage in tests is extremely common.

Sure, but tests / playing around is IMO not important enough to complicate the type.

Either way I think the right approach is to promote `cats.Eval` as being the (guaranteed) synchronous `IO`-like type.

Speaking of which, how do we make `cats.Eval` implement `Sync`? Should we place a `Sync[Eval]` instance in the `Sync` companion object?

**Comment by djspiewak** [comment]
Going to close this for now.  I think it's interesting discussion, but has mostly run its course.  We can explore some of the ideas herein (e.g. typed specializations of `IO`) in other tickets and PRs.
